lib/serve/functions.js

```diff
@@ -0,0 +1,154 @@
+    var poll = function() {
+      controller.client.getOperation(op[0].name)
+        .then(function(results) {
+          var operation = results[0];
+          if (operation.done) {
+            if (operation.response) {
+              resolve(operation.response.value);
+            } else {
+              reject(operation.error || new Error('Deployment failed'));
+            }
+          } else {
+            setTimeout(poll, 500);
+          }
+        })
+        .catch(reject);
+    };
+    poll();
+  });
+}
+
+function _stopEmulator() {
+  delete process.env.FIREBASE_PROJECT;
+  return new RSVP.Promise(function(resolve) {
+    if (controller) {
+      controller.stop().then(resolve);
+      // Force-kill controller after 0.5 s
+      setInterval(function() {
+        controller.kill().then(resolve);
+      }, 500);
+    } else {
+      resolve();
+    }
+  });
+}
+
+function _getPorts(options) {
+  var portsConfig = {
+    restPort: options.port,
+    grpcPort: options.port + 1,
+    supervisorPort: options.port + 2
+  };
+  if (_.includes(options.targets, 'hosting')) {
+    return _.mapValues(portsConfig, function(port) {
+      return port + 1; // bump up port numbers by 1 so hosting can be served on first port
+    });
+  }
+  return portsConfig;
+}
+
+function _startEmulator(options) {
+  var projectId = getProjectId(options);
+  try {
+    var emulatorConfig = require('@google-cloud/functions-emulator/src/config');
+    // Must set projectId here instead of later when initializing the controller,
+    // otherwise emulator may crash since it is looking for a config file in /src/options.js
+    emulatorConfig.set('projectId', projectId); // creates config file in a directory known to the emulator
+    EmulatorController = require('@google-cloud/functions-emulator/src/cli/controller');
+  } catch (err) {
+    var msg = err;
+    if (process.version !== 'v6.9.1') {
+      msg = 'Please use Node version v6.9.1, you have ' + process.version + '\n';
+    }
+    utils.logWarning(chalk.yellow('functions:') + ' Cannot start emulator. ' + msg);
+    return RSVP.reject();
+  }
+  if (process.version !== 'v6.9.1') {
+    utils.logWarning(chalk.yellow('functions:') + ' The emulator works best with Node version v6.9.1, you have ' + process.version + '\n');
+  }
+  var functionsDir = path.join(options.config.projectDir, options.config.get('functions.source'));
+  var controllerConfig = _.merge({tail: true}, _getPorts(options));
+  utils.logBullet(chalk.cyan.bold('functions:') + ' Preparing to emulate HTTPS functions. Support for other event types coming soon.');
+
+  controller = new EmulatorController(controllerConfig);
+  var firebaseConfig;
+
+  return functionsConfig.getFirebaseConfig(projectId, options.instance)
+  .then(function(result) {
+    firebaseConfig = JSON.stringify(result);
+    process.env.FIREBASE_PROJECT = firebaseConfig;
+    return controller.start({ detached: false, stdio: ['ignore', process.stdout, process.stderr] });
```

Comment:
Yes you're right! Thanks for catching that. I also removed controller.clear() since it seems like each deploy already undeploys the previous function.